ltiauthenticator/lti11/auth.py

```diff
@@ -1,70 +1,84 @@
+    JupyterHub LTI 1.1 Authenticator which extends the ltiauthenticator.LTIAuthenticator class.
+    Messages sent to this authenticator are sent from a tool consumer (TC), such as
+    an LMS. JupyterHub, as the authenticator, works as the tool provider (TP), also
+    known as the external tool.
+
+    The LTIAuthenticator base class defines the consumers, defined as 1 or (n) consumer key
+    and shared secret k/v's to verify requests from their tool consumer.
     """
 
     auto_login = True
-    login_service = "LTI"
+    login_service = "LTI 1.1"
 
     consumers = Dict(
         {},
         config=True,
         help="""
         A dict of consumer keys mapped to consumer secrets for those keys.
-
         Allows multiple consumers to securely send users to this JupyterHub
         instance.
         """,
     )
 
-    def get_handlers(self, app):
+    def get_handlers(self, app: JupyterHub) -> BaseHandler:
         return [("/lti/launch", LTI11AuthenticateHandler)]
 
-    @gen.coroutine
-    def authenticate(self, handler, data) -> dict:
-        # FIXME: Run a process that cleans up old nonces every other minute
+    def login_url(self, base_url):
+        return url_path_join(base_url, "/lti/launch")
+
+    async def authenticate(  # noqa: C901
+        self, handler: BaseHandler, data: dict = None
+    ) -> dict:  # noqa: C901
+        """
+        LTI 1.1 Authenticator. One or more consumer keys/values must be set in the jupyterhub config with the
+        LTI11Authenticator.consumers dict.
+
+        Args:
+            handler: JupyterHub's Authenticator handler object. For LTI 1.1 requests, the handler is
+              an instance of LTIAuthenticateHandler.
+            data: optional data object
+
+        Returns:
+            Authentication dictionary
+
+        Raises:
+            HTTPError if the required values are not in the request
+        """
         validator = LTI11LaunchValidator(self.consumers)
 
-        args = {}
-        for k, values in handler.request.body_arguments.items():
-            args[k] = (
-                values[0].decode() if len(values) == 1 else [v.decode() for v in values]
-            )
-
-        # handle multiple layers of proxied protocol (comma separated) and take the outermost
-        # value (first from the list)
-        if "x-forwarded-proto" in handler.request.headers:
-            # x-forwarded-proto might contain comma delimited values
-            # left-most value is the one sent by original client
-            hops = [
-                h.strip()
-                for h in handler.request.headers["x-forwarded-proto"].split(",")
-            ]
-            protocol = hops[0]
-        else:
-            protocol = handler.request.protocol
-
-        launch_url = protocol + "://" + handler.request.host + handler.request.uri
+        self.log.debug(
+            "Original arguments received in request: %s" % handler.request.arguments
```

Comment:
Will these be sensitive?

Reply:
Not that I am aware of. They could be sensitive if the integration with the LMS is set to private, in which case the username is fetched from the `user_id` which is an opaque identifier and not considered personally identifiable information (PII). But in these cases the arguments that would contain PII values are empty. The other arguments are related to context (courses) or assignments/modules (resources).

Reply:
Given that it's debug logging I think it's ok. We already log user_ids elsewhere in JupyterHub.